pkg/workflow/workflow.go

```diff
@@ -270,79 +290,88 @@ func (w *workflow) Trace() error {
+			if step.Type == wfTypes.WorkflowStepTypeSuspend || step.Type == wfTypes.WorkflowStepTypeStepGroup {
+				min = handleSuspendBackoffTime(step, stepStatus[step.Name], min)
+			}
+			for _, sub := range step.SubSteps {
+				if sub.Type == wfTypes.WorkflowStepTypeSuspend {
+					min = handleSuspendBackoffTime(oamcore.WorkflowStep{
+						Name:       sub.Name,
+						Type:       sub.Type,
+						Timeout:    sub.Timeout,
+						Properties: sub.Properties,
+					}, stepStatus[sub.Name], min)
+				}
+			}
+		}
 	}
-
-	return time.Second
+	if min == max {
+		return 0
+	}
+	return min
 }
 
-func (w *workflow) HandleSuspendWait(ctx monitorContext.Context) (doWaiting bool, durationWaiting time.Duration, errRet error) {
-	ctx.Info("handle suspend wait")
-	for i, stepStatus := range w.app.Status.Workflow.Steps {
-		if !w.isWaitSuspendStep(stepStatus) {
-			continue
-		}
-
-		step := w.getWorkflowStepByName(stepStatus.Name)
-		if step.Name == "" {
-			errRet = fmt.Errorf("failed to get workflow step by name: %s", stepStatus.Name)
-			return
-		}
-
-		d, wd, err := wfTasks.GetSuspendStepDurationWaiting(step)
-		if err != nil {
-			ctx.Error(err, "failed to get suspend step wait duration")
-			errRet = err
-			return
-		}
-
-		if d {
-			doWaiting = d
-			remainingDuration := wd - time.Since(stepStatus.FirstExecuteTime.Time)
-			if remainingDuration <= 0 {
-				w.app.Status.Workflow.Steps[i].Phase = common.WorkflowStepPhaseSucceeded
+func handleSuspendBackoffTime(step oamcore.WorkflowStep, status common.StepStatus, min time.Duration) time.Duration {
+	if status.Phase == common.WorkflowStepPhaseRunning {
+		if step.Timeout != "" {
+			duration, err := time.ParseDuration(step.Timeout)
+			if err != nil {
+				return min
 			}
-
-			if remainingDuration > 0 && (durationWaiting > remainingDuration || durationWaiting <= 0) {
-				suspendState := fmt.Sprintf("durationWaiting(%s)", wd.String())
-				if w.app.Status.Workflow.SuspendState != suspendState {
-					w.app.Status.Workflow.SuspendState = suspendState
+			timeout := status.FirstExecuteTime.Add(duration)
+			if time.Now().Before(timeout) {
+				d := time.Until(timeout)
+				if duration < min {
+					min = d
 				}
-				durationWaiting = remainingDuration
 			}
 		}
 
-		if !w.dagMode {
-			return
+		d, err := wfTasks.GetSuspendStepDurationWaiting(step)
+		if err != nil {
+			return min
+		}
+		if d != 0 && d < min {
+			min = d
 		}
-	}
-
-	return doWaiting, durationWaiting, errRet
-}
 
-func (w *workflow) isWaitSuspendStep(status common.WorkflowStepStatus) bool {
-	return status.Type == wfTypes.WorkflowStepTypeSuspend && status.Phase == common.WorkflowStepPhaseRunning
+	}
+	return min
 }
 
-func (w *workflow) getWorkflowStepByName(name string) oamcore.WorkflowStep {
-	for _, s := range w.app.Spec.Workflow.Steps {
-		if s.Name == name {
-			return s
+func (w *workflow) GetBackoffWaitTime() time.Duration {
+	nextTime, ok := w.wfCtx.GetValueInMemory(wfTypes.ContextKeyNextExecuteTime)
```

Comment:
what will happend if the controller crash and restared?

Reply:
We'll lose the `NextExecuteTime` in memory so the backoff time is re-calculated from the initial value `1`.